tests/vmi_lifecycle_test.go

```diff
@@ -1613,12 +1613,14 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 				}, 30, 0.5).Should(ContainSubstring(fmt.Sprintf("Signaled graceful shutdown for %s", vmi.GetObjectMeta().GetName())), "Should log graceful shutdown")
 
 				// Verify VirtualMachineInstance is killed after grace period expires
-				By("Checking that the VirtualMachineInstance does not exist after grace period")
-				Eventually(func() string {
-					data, err := logsQuery.DoRaw(context.Background())
-					Expect(err).ToNot(HaveOccurred(), "Should get logs")
-					return string(data)
-				}, 30, 0.5).Should(ContainSubstring(fmt.Sprintf("Grace period expired, killing deleted VirtualMachineInstance %s", vmi.GetObjectMeta().GetName())), "Should log graceful kill")
+				if !libvmifact.IsS390X() { //Exclude this for the S390x architecture, as graceful shutdown is sufficient.
```

Comment:
Can you remove this? It doesn't right to exclude s390x from this final check, or do you have any explanation why you do this?

Reply:
Ideally, we shouldn't need to send a kill signal if the shutdown process succeeds.

On x86 Architecture:

In the minimal Alpine Linux image, the acpid daemon is not installed in user space. While QEMU sends an ACPI signal to the kernel, the absence of the acpid daemon means that user-space processes do not receive the shutdown event, preventing a complete graceful shutdown at the user-space level. However, kernel-level operations like file synchronization may still proceed.

output vmi stages for reference:

root@anand-nekkunti-ubuntux86:~# kubectl get vmi -A -w
NAMESPACE                NAME            AGE   PHASE     IP    NODENAME   READY
kubevirt-test-default1   testvmi-txhzt   1s    Pending
kubevirt-test-default1   testvmi-txhzt   1s    Pending
kubevirt-test-default1   testvmi-txhzt   1s    Pending                    False
kubevirt-test-default1   testvmi-txhzt   1s    Scheduling                    False
kubevirt-test-default1   testvmi-txhzt   1s    Scheduling                    False
kubevirt-test-default1   testvmi-txhzt   6s    Scheduled          node01     False
kubevirt-test-default1   testvmi-txhzt   8s    Scheduled          node01     False
kubevirt-test-default1   testvmi-txhzt   8s    Running      10.244.196.154   node01     False
kubevirt-test-default1   testvmi-txhzt   8s    Running      10.244.196.154   node01     True
kubevirt-test-default1   testvmi-txhzt   29s   Running      10.244.196.154   node01     True
kubevirt-test-default1   testvmi-txhzt   29s   Running      10.244.196.154   node01     False
kubevirt-test-default1   testvmi-txhzt   89s   **Failed**       10.244.196.154   node01     False
kubevirt-test-default1   testvmi-txhzt   90s   **Failed**       10.244.196.154   node01     False


As a result, the VM running Alpine on x86 enters a failed state even after receiving the graceful shutdown signal, so we end up needing to send a kill signal in this condition.

When I replaced Alpine with a Fedora VM, which has the QEMU guest agent running, the shutdown is initiated via this agent for both x86 and s390x architectures. In this case, the VM shuts down successfully, eliminating the need to send a kill signal.

It seems this test targets the worst-case scenario where ACPI is available in the kernel but the acpid daemon is not present. In such cases on x86, we end up sending both a graceful shutdown signal and a kill signal after grace priod expired.

On s390x Architecture:

The s390x architecture (used by IBM systems) doesn't support ACPI. Instead, it uses the QEMU guest agent or architecture-specific logic to shut down, which consistently succeeds. Therefore, we don't need to send a kill signal for s390x unless the VM enters a failed state.

Reply:
Gotcha, but to me it seems like the test itself is buggy, we should be testing a graceful shutdown without accounting for the VM to not have acpid, of course this PR is not to blame, the code was already like this.

I'd suggest switching to Fedora (which does have the acpi daemon) and remove that last check entirely for all arches, in the end as I said we want to test graceful shutdown, not that the VM gets forcibly killed when the grace period has expired, there should be another test that checks that.

Do you agree?

Reply:
@nekkunti I'd change the test to be like this

```go
Context("with grace period greater than 0", func() {
			DescribeTable("the VM should be", func(vmi *v1.VirtualMachineInstance, shouldBeKilled bool) {
				nodes := libnode.GetAllSchedulableNodes(kubevirt.Client())
				Expect(nodes.Items).ToNot(BeEmpty(), "There should be some compute node")
				node := nodes.Items[0].Name

				virtHandlerPod, err := libnode.GetVirtHandlerPod(kubevirt.Client(), node)
				Expect(err).ToNot(HaveOccurred(), "Should get virthandler for node")

				handlerName := virtHandlerPod.GetObjectMeta().GetName()
				handlerNamespace := virtHandlerPod.GetObjectMeta().GetNamespace()
				seconds := int64(120)
				logsQuery := kubevirt.Client().CoreV1().Pods(handlerNamespace).GetLogs(handlerName, &k8sv1.PodLogOptions{SinceSeconds: &seconds, Container: "virt-handler"})

				By("Setting a VirtualMachineInstance termination grace period to 5")
				var gracePeriod int64
				gracePeriod = int64(5)
				// Give the VirtualMachineInstance a custom grace period
				vmi.Spec.TerminationGracePeriodSeconds = &gracePeriod
				// Make sure we schedule the VirtualMachineInstance to master
				vmi.Spec.NodeSelector = map[string]string{k8sv1.LabelHostname: node}

				By("Creating the VirtualMachineInstance")
				vmi = libvmops.RunVMIAndExpectLaunch(vmi, startupTimeout)

				// Delete the VirtualMachineInstance and wait for the confirmation of the delete
				By("Deleting the VirtualMachineInstance")
				Expect(kubevirt.Client().VirtualMachineInstance(vmi.Namespace).Delete(context.Background(), vmi.Name, metav1.DeleteOptions{})).To(Succeed(), "Should delete VMI gracefully")
				ctx, cancel := context.WithCancel(context.Background())
				defer cancel()
				event := watcher.New(vmi).SinceWatchedObjectResourceVersion().Timeout(75*time.Second).WaitFor(ctx, watcher.NormalEvent, v1.Deleted)
				Expect(event).ToNot(BeNil(), "There should be a delete event")

				// Check if the graceful shutdown was logged
				By("Checking that virt-handler logs VirtualMachineInstance graceful shutdown")
				Eventually(func() string {
					data, err := logsQuery.DoRaw(context.Background())
					Expect(err).ToNot(HaveOccurred(), "Should get the logs")
					return string(data)
				}, 30, 0.5).Should(ContainSubstring(fmt.Sprintf("Signaled graceful shutdown for %s", vmi.GetObjectMeta().GetName())), "Should log graceful shutdown")

				if shouldBeKilled {
					// Verify VirtualMachineInstance is killed after grace period expires
					By("Checking that the VirtualMachineInstance does not exist after grace period")
					Eventually(func() string {
						data, err := logsQuery.DoRaw(context.Background())
						Expect(err).ToNot(HaveOccurred(), "Should get logs")
						return string(data)
					}, 30, 0.5).Should(ContainSubstring(fmt.Sprintf("Grace period expired, killing deleted VirtualMachineInstance %s", vmi.GetObjectMeta().GetName())), "Should log graceful kill")
				}
			})
		},
			Entry("[test_id:1655] gracefully shutdown", decorators.WgS390x, libvmifact.NewFedora(), false),
			Entry("killed after the grace period expires", libvmifact.NewCirros(), true),
		)
```

Reply:
agree ., got it 
let me see , any other better solution else i will do this way.

Thanks @acardace